Studio: Add UI to import new site during Onboarding and add site modal

katinthehatsite opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

Closes https://github.com/Automattic/dotcom-forge/issues/8159
Closes https://github.com/Automattic/dotcom-forge/issues/8158
## Proposed Changes

This PR adds the new import site UI to onboarding screen and to `Add site` modal:
* it adds the functionality to drag and drop the import file
* it adds the functionality to open the file manager and select the file for import

## Testing Instructions

**Testing onboarding screen**

Before starting the testing:
* Add a specific theme to one of your local sites in Studio and make sure it is installed
* Navigate to the Import/Export tab
* Export that site (not the database)

Once you have the backup file:
* Delete all local sites
* Pull the changes from this branch
* Start the app with `STUDIO_IMPORT_EXPORT=true npm start`
* Observe that you can now see a new field on the onboarding screen that says `Import a backup`:

<img width="569" alt="Screenshot 2024-07-25 at 8 22 47 AM" src="https://github.com/user-attachments/assets/4a100594-98fc-4ac1-94d8-15dd664cc3e6">

* Click on `Advanced Settings` button and confirm that you can see the `Local path` appear
* Click again on `Advanced Settings` button  and confirm that the field with `Local path` closes
* Select a backup file to import (that one you previously exported)
* Drag it into the `Add your first site` area (the white part of the onboarding)
* Observe the white overlay and the blue download icon appearing while you are dragging the file
* Observe that once you drag the file, its name is filled into the `Import the backup` field and the trash icon appears:

<img width="532" alt="Screenshot 2024-07-25 at 8 27 57 AM" src="https://github.com/user-attachments/assets/511c6356-a5fe-40aa-9a42-c6436682c906">

* Click on the trash icon and confirm that it clears the input field
* Try dragging an invalid import file e.g. `.png` - confirm that you can't drag it into the form and you see the error
* Select a backup file to import (that one you previously exported)
* Drag it into the `Add your first site` area (the white part of the onboarding)
* This time, proceed with `Add site` button
* Observe that you see the loading screen
* Observe that once the site is created, you get the `Import completed` notification
* Open the new site and navigate to `Themes`
* Confirm that you can see the themes that were present in the export file on the new site
* Confirm that the site works as expected

**Testing with the file manager**

* Delete the created site to get back to the onboarding screen
* Click on the folder icon or on the input field for `Import a backup`
* Observe that it opens the file manager and allows you to select only `tar.gz`, `.zip`, `.sql`
* Select the backup file to import 
* Observe that the input field is populating with the file name
* Proceed with the `Add site` button
* Observe that you see the loading screen
* Observe that once the site is created, you get the `Import completed` notification
* Open the new site and navigate to `Themes`
* Confirm that you can see the themes that were present in the export file on the new site
* Confirm that the site works as expected

**Testing Add site modal**

* Click on the `Add site` button in the sidebar to open the `Add site` modal
* Observe that you can see the new input field to add the backup file on the modal
* Select a backup file to import (that one you previously exported)
* Drag it into the modal area
* Observe that the design with the white overlay and the blue icon appears while you are dragging the file
* Observe that once you drag the file, its name is filled into the `Import the backup` field and the trash icon appears
* This time, proceed with `Add site` button
* Observe that you see the loading screen
* Observe that once the site is created, you get the `Import completed` notification
* Open the new site and navigate to `Themes`
* Confirm that you can see the themes that were present in the export file on the new site
* Confirm that the site works as expected
* Repeat the steps above but instead of dragging the file, test it with the file manager

**Check Edit Site form**

(Added these testing steps as the site form is sharing the same component with `Add site` modal and onboarding

* Navigate to the `Settings` tab for a site that is already created
* Click on `Edit` by the `Site name`
* Observe that you can only see the field to edit the site name and nothing else

**Test feature flag**
* Pull the changes from this branch
* Start the app with `npm start`
* Confirm that you see the UI that is currently available in the app
* Quit the app and start the app with `STUDIO_IMPORT_EXPORT=true npm start`
* Confirm that you see the new UI in the onboarding screen and `Add site` modal
## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Have you checked for TypeScript, React or other console errors?

Commit:
Add new field for uploading an import

Commit:
Add placeholder to form

Commit:
Add drag and drop logic

Commit:
Add dragging over view

Commit:
Add icon for when the file name is set

Commit:
Add functionality to clear the file name

Commit:
Move local path to advanced settings

Commit:
Fix type error

Commit:
Fix the trash icon button

Commit:
Fix label overlap

Commit:
Open file editor from the form

Commit:
Update input field with selected file

Commit:
Merge trunk

Commit:
Integrate import logic

Commit:
Fix type errors on site form

Commit:
Add drag and drop to the add site form

Commit:
Clear file when the modal is closing

Commit:
Add style for dragging state on modal

Commit:
Ensure that the Edit form remains unchanged

Commit:
Ensure that the loading state on import tab does not break

Commit:
Cleanup console log

Commit:
Cleanup comments

Commit:
Refactor trash button

katinthehatsite commented on src/components/site-content-tabs.tsx:
```diff
@@ -24,7 +24,7 @@ export function SiteContentTabs() {
 	}
 
-	if ( selectedSite?.isAddingSite ) {
+	if ( selectedSite?.isAddingSite || selectedSite?.importState === 'new-site-importing' ) {
```
This looks strange but it is necessary to be able to differentiate if the site is importing for the first time or not. This differentiation helps us to ensure that the loading state does not break on the `Import/ Export` tab and we can use the distinct loading state for the first time import which is what we have in the design.

Commit:
Ensure that invalid filetype can't be accepted

katinthehatsite commented:
FYI, ideally we should combine some import logic from this PR https://github.com/Automattic/studio/pull/369/files and this current PR and refactor some of it into the hook. Since this PR is already quite large, I did not do it here but we can handle it as a separate issue if we decide to do so.

I am fixing tests but I will request the review in the meantime to start getting feedback.

Commit:
Move file types to a constant

Commit:
Move drag and drop overlay into a separate component

Commit:
Refactor errors into one component

Commit:
Fix failing tests

sejas commented on src/hooks/use-drag-and-drop-file.ts:
```diff
@@ -36,10 +36,26 @@ export function useDragAndDropFile< T extends HTMLElement >( {
+			setIsDraggingOver( false );
+		};
+
+		const node = dropRef.current;
```
What do you think about using dropNode, dropZone, dropElement instead of just node?
We could declare it at the beginning of the useEffect, and use it in the rest of the function.

```suggestion
		const dropNode = dropRef.current;
```

katinthehatsite replied:
Done in https://github.com/Automattic/studio/pull/378/commits/cdcac1b629a7b1b877725714021416e0368fb347
As mentioned in another comment, I also removed the `dragend`

sejas commented on src/hooks/use-site-details.tsx:
```diff
@@ -165,7 +169,7 @@ export function SiteDetailsProvider( { children }: SiteDetailsProviderProps ) {
 
 	const createSite = useCallback(
-		async ( path: string, siteName?: string ) => {
+		async ( path: string, siteName?: string, isImportingNewSite?: boolean ) => {
```
Would be interesting initialize the parameter to false?

```suggestion
		async ( path: string, siteName?: string, isImportingNewSite: boolean = false ) => {
```

katinthehatsite replied:
Yeah makes sense, updated in https://github.com/Automattic/studio/pull/378/commits/11a8dd8a9d170230444a94b2c9a52d22449a8185

sejas commented on src/components/add-site.tsx:
```diff
@@ -89,13 +98,33 @@ export default function AddSite( { className }: AddSiteProps ) {
+		async ( file: File ) => {
+			setFileForImport( file );
+		},
+		[ setFileForImport ]
```
What's the benefit of this function? Could we use setFileForImport directly?
If you want to rename that funciton, you could make a simple assignment `const handleImportFile = setFileForImport` or rename it when you deconstruct it from `useAddSite`. But I think it would be necessary in this case.

```suggestion
```

katinthehatsite replied:
I was initially going to remove this but I found an edge case where this is useful in its updated form - I made some changes to the function here: https://github.com/Automattic/studio/pull/378/commits/c6f1ef89624735853860a64cf0cc314334c69ddb

The scenario where I found we need this is the following:
* if the user tries to drag and drop an invalid file type, the form displays an error message
* if then instead of drag and drop, they click on the input field and open the file finder and select a file, the error state needs to be updated
* otherwise, the error message will be hanging even though the file would be correct

I added `setFileError( '' )` to that function to ensure that this case remains covered. Let me know what you think!

sejas commented on src/components/add-site.tsx:
```diff
@@ -113,29 +142,36 @@ export default function AddSite( { className }: AddSiteProps ) {
+							doesPathContainWordPress={ doesPathContainWordPress }
+							fileForImport={ fileForImport }
+							setFileForImport={ setFileForImport }
+							onFileSelected={ handleImportFile }
```
Following the comment above, we could use directly `setFileForImport`:

```suggestion
							onFileSelected={ setFileForImport }
```

katinthehatsite replied:
See the comment in https://github.com/Automattic/studio/pull/378#discussion_r1691466673 - I will mark this bit as resolved for now based on my explanation.

sejas commented on src/constants.ts:
```diff
@@ -17,6 +17,15 @@ export const FEATURE_REQUEST_URL =
 
+//Import file constants
+
+export const ACCEPTED_FILE_TYPES = [
```
Nit: what do you think about renaming the variable itself and removing the comment?
```suggestion
//Import file constants

export const ACCEPTED_IMPORT_FILE_TYPES = [
```

Or maybe ACCEPTED_BACKUP_FILE_TYPES

katinthehatsite replied:
Let's use `ACCEPTED_IMPORT_FILE_TYPES` - I think this verbiage is clear for the purpose of it 👍  
Updated in https://github.com/Automattic/studio/pull/378/commits/c3baca777a7beaf1623779fbb44fc5aa6a0bc72d

sejas commented on src/hooks/use-add-site.ts:
```diff
@@ -110,20 +114,23 @@ export function useAddSite() {
+			siteWithPathAlreadyExists,
 			sitePath,
 			proposedSitePath,
+			__,
+			error,
+			siteName,
+			doesPathContainWordPress,
```
We could let these dependencies as in the previous order to avoid introducing changes not needed.

katinthehatsite replied:
Resolved in https://github.com/Automattic/studio/pull/378/commits/f03498c76e9f5878530e72a72ae02a29dc4d21d1

sejas commented on src/hooks/use-drag-and-drop-file.ts:
```diff
@@ -36,10 +36,26 @@ export function useDragAndDropFile< T extends HTMLElement >( {
-		dropRef.current.addEventListener( 'dragleave', handleDragLeave );
-		dropRef.current.addEventListener( 'drop', handleDrop );
 
 		function cleanup() {
 			if ( ! dropRef.current ) {
```
If we return the remove listeners directly, then we may want to remove the cleanup function and the second return.

katinthehatsite replied:
I kept this original logic as after some testing, it worked well and removed further modifications in https://github.com/Automattic/studio/pull/378/commits/cdcac1b629a7b1b877725714021416e0368fb347

sejas commented on src/components/site-form.tsx:
```diff
@@ -106,29 +223,77 @@ export const SiteForm = ( {
+							<FormImportComponent
+								placeholder={ __( 'Select or drop a file' ) }
+								value={ fileForImport }
+								onChange={ ( file ) => setFileForImport( file ) }
```
You could pass directly the function.

```suggestion
								onChange={ setFileForImport }
```

katinthehatsite replied:
Updated in https://github.com/Automattic/studio/pull/378/commits/5226cac20e816c871c0dfeeb88f527537a64f1b9

sejas commented on src/components/site-form.tsx:
```diff
@@ -106,29 +223,77 @@ export const SiteForm = ( {
+								<div className="text-[13px] leading-[16px] ml-2">{ __( 'Advanced settings' ) }</div>
+							</Button>
+						</div>
+						{ onSelectPath && isAdvancedSettingsVisible && (
```
Should this `onSelectPath` check be wrapping around the "advanced settings" ? I want to avoid the case where we forget to pass `onSelectPath` and then we have an empty expandable "advanced view"
```suggestion
						{ isAdvancedSettingsVisible && (
```

sejas replied:
Not a strong opinion, since we this behaviour will change and make more sense when we add the WP and PHP versions inside the advanced view.

katinthehatsite replied:
you are right, good catch. I updated this in https://github.com/Automattic/studio/pull/378/commits/8e36959b9173999606b0d4f43acf5ff3facf5f1d

I think we will end up refactoring this in the future probably anyway but it is a good idea to take care of this at the moment

sejas commented on src/hooks/use-drag-and-drop-file.ts:
```diff
@@ -36,10 +36,26 @@ export function useDragAndDropFile< T extends HTMLElement >( {
+		node.addEventListener( 'dragover', handleDragOver );
+		node.addEventListener( 'dragleave', handleDragLeave );
+		node.addEventListener( 'drop', handleDrop );
+		node.addEventListener( 'dragend', handleDragEnd );
```
Out of curiosity, why do we need to listen the dragend event? 
I thought it was to support the `esc` key and abort the drag and drop, but it seems to work also without listening this event.

katinthehatsite replied:
I will double-check again but as far as I remember, I could not make the drag and drop work over the whole area without using `dragend` event and the `isDragging` was just hanging in there as `true` creating a weird UI.

katinthehatsite replied:
I took and a look and it seems to work now without `dragend`. Not completely sure what happened but it was not working before 😅 I did use the hook though before your PR was merged into trunk and it is possible that some updates were made after I used the hook that solved the issue. I removed the `dragend` in https://github.com/Automattic/studio/pull/378/commits/cdcac1b629a7b1b877725714021416e0368fb347

sejas commented on src/components/onboarding.tsx:
```diff
@@ -89,6 +107,13 @@ export default function Onboarding() {
+			setFileForImport( file );
+		},
+		[ setFileForImport ]
+	);
```
Similar to add-site.tsx, you could use directly setFileForImport without creating a intermediary function.
```suggestion
	const handleImportFile = useCallback(
		async ( file: File ) => {
			setFileForImport( file );
		},
		[ setFileForImport ]
	);
```

katinthehatsite replied:
See the comment in https://github.com/Automattic/studio/pull/378#discussion_r1691466673 - I will mark this bit as resolved for now based on my explanation.

sejas reviewed (commented):
@katinthehatsite , great work!
I tested it and didn't find any bug or issue.
I left a few improvements/suggestions.
Feel free to apply those that resonate more with you.

I've noticed that the "import" icon should be bigger. Our icon has a "padding" and that's why it looks smaller.

Code:
<img width="1012" alt="Screenshot 2024-07-25 at 13 27 09" src="https://github.com/user-attachments/assets/138c960b-e76d-4c6b-bbf3-ac5e1f1f1f0c">

Figma:

<img width="1092" alt="Screenshot 2024-07-25 at 13 37 15" src="https://github.com/user-attachments/assets/091f5ca5-9679-4fbe-bee8-ef8169069083">

See here the actual size of the icon, targeting the path element:
<img width="243" alt="Screenshot 2024-07-25 at 13 28 12" src="https://github.com/user-attachments/assets/73a61aa5-c75c-4d23-b904-43a7de145503">

Commit:
Handle the error that remains

Commit:
Initialize isImportingNewSite to false

Commit:
Rename constant for file types

Commit:
Set file for import directly

wojtekn commented on src/components/onboarding.tsx:
```diff
@@ -98,7 +124,8 @@ export default function Onboarding() {
 
-			<div className="w-1/2 bg-white p-[50px] flex flex-col">
+			<div className="w-1/2 bg-white p-[50px] flex flex-col relative" ref={ dropRef }>
+				{ isDraggingOver && <DragAndDropOverlay /> }
```
Can we account for the feature flag in the new import logic during onboarding?

wojtekn replied:
It seems in the 'Add site' form we also miss the feature flag.

katinthehatsite replied:
Yeah, I will add it in these two places 👍

katinthehatsite replied:
Done in https://github.com/Automattic/studio/pull/378/commits/f0f7a450a7b6faa5e466656f87829f50477e6b1f

I updated the testing steps in the PR to reflect these changes

wojtekn reviewed (commented):
Looks good in general and works smoothly. I have a couple of comments related to the onboarding UI:
- The 'Add site' button should be aligned to right
- Should we animate expanding and collapsing advanced options?
- Designs show notifications displayed for newly added sites. Now we show that, but only the one for the import. Could we add that notification, too, or at least ensure a separate issue is added to cover that?

<img width="1012" alt="Screenshot 2024-07-26 at 09 01 20" src="https://github.com/user-attachments/assets/51892291-f4e1-4815-9988-3be3f8bf99dd">

Commit:
Align Add site button to the right

Commit:
Fix icon size

katinthehatsite commented:
Updates:

* Fixed icon size on the onboarding screen: https://github.com/Automattic/studio/pull/378/commits/a3af107ea9173d740cd467302d4e4743bcab2178
* Fixed `Add site alignment`: https://github.com/Automattic/studio/pull/378/commits/a047003e8eb44c9e181508c91a5306a31b19d3e6

Commit:
Wrap condition to render onSelectPath

Commit:
Cleanup drag and drop hook

Commit:
Restore the order of dependecies

Commit:
Restore the order of dependecies

Commit:
Drag and drop file cleanup 2

Commit:
Wrap new UI in the feature flag

Commit:
Fix failing tests

katinthehatsite commented:
Thanks @wojtekn !

> The 'Add site' button should be aligned to right

* Fixed `Add site alignment`: https://github.com/Automattic/studio/pull/378/commits/a047003e8eb44c9e181508c91a5306a31b19d3e6

> Should we animate expanding and collapsing advanced options?

I already thought about this when initially adding the UI to make the jump a bit more smooth perhaps. Anything specific you are looking for with this animation?

> Designs show notifications displayed for newly added sites. Now we show that, but only the one for the import. Could we add that notification, too, or at least ensure a separate issue is added to cover that?

Let's open a separate issue. I will create one and handle it there as this PR is getting quite large already 👍 
Opened: https://github.com/Automattic/dotcom-forge/issues/8461

wojtekn commented:
> I already thought about this when initially adding the UI to make the jump a bit more smooth perhaps. Anything specific you are looking for with this animation?

I think the transition between both states should be smoother, and not instant. We can move it to the separate issue, though.

sejas commented:
In case it can be used as reference, this is the only animation I've seen prototyped:
p9Jlb4-cF2-p2#comment-12114


https://github.com/user-attachments/assets/647382cd-d417-4c60-b3cb-bfad758ce269

sejas reviewed (approved):
@katinthehatsite , great work.

I've tested it with and without feature flag and it works as expected. Thank you!

With Feature flag
https://github.com/user-attachments/assets/f45206d5-359a-47da-a880-78a1f5de7581

Without feature flag
<img width="1012" alt="Screenshot 2024-07-26 at 11 28 44" src="https://github.com/user-attachments/assets/30aa5c31-6946-463c-914f-8eb08c8bcb0e">
<img width="1012" alt="Screenshot 2024-07-26 at 11 28 57" src="https://github.com/user-attachments/assets/fe63d42d-f8cd-423a-8b41-88cc062308b0">

katinthehatsite commented:
I opened a new issue to add some animation here: https://github.com/Automattic/dotcom-forge/issues/8468
